File: android/src/main/java/org/ligi/passandroid/ui/edit/FieldsEditFragment.java
package org.ligi.passandroid.ui.edit;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.text.Editable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageButton;
import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import java.util.List;
import org.ligi.axt.simplifications.SimpleTextWatcher;
import org.ligi.passandroid.R;
import org.ligi.passandroid.model.pass.PassField;

public class FieldsEditFragment extends PassandroidFragment {

    private final static String ARGUMENT_KEY = "KEY";
    private boolean isEditingHiddenFields;

    public static FieldsEditFragment create(boolean primary) {
        final FieldsEditFragment fieldsEditFragment = new FieldsEditFragment();
        final Bundle args = new Bundle();
        args.putBoolean(ARGUMENT_KEY, primary);
        fieldsEditFragment.setArguments(args);
        return fieldsEditFragment;
    }

    @Bind(R.id.fields_container)
    ViewGroup viewGroup;

    @Bind(R.id.add_field)
    Button addFieldButton;

    @OnClick(R.id.add_field)
    void onAddField() {
        final PassField passField = new PassField(null, null, null, isEditingHiddenFields);
        addField(passField);
        getPass().getFields().add(passField);
    }

    private LayoutInflater inflater;

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        this.inflater = inflater;
        final View inflate = inflater.inflate(R.layout.edit_fields, container, false);

        ButterKnife.bind(this, inflate);

        isEditingHiddenFields = getArguments().getBoolean(ARGUMENT_KEY);

        if (isEditingHiddenFields) {
            addFieldButton.setText(R.string.add_back_fields);
        } else {
            addFieldButton.setText(R.string.add_front_field);
        }
        for (final PassField passField : getPass().getFields()) {
            if (passField.getHide() == isEditingHiddenFields) {
                addField(passField);
            }
        }

        return inflate;
    }

    public void addField(final PassField passField) {
        final ViewGroup child = (ViewGroup) inflater.inflate(R.layout.edit_field, viewGroup, false);
        new FieldView(child).apply(passField, getPass().getFields());
        viewGroup.addView(child);
    }

    class FieldView {
        @Bind(R.id.label_field_edit)
        TextInputEditText labelEdit;

        @Bind(R.id.value_field_edit)
        TextInputEditText valueEdit;

        @Bind(R.id.delete_button)
        ImageButton deleteButton;

        FieldView(final ViewGroup container) {
            ButterKnife.bind(this, container);
        }

        public void apply(final PassField passField, final List<PassField> fields) {
            labelEdit.setText(passField.getLabel());
            valueEdit.setText(passField.getValue());

            deleteButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(final View v) {
                    ((View) (v.getParent())).setVisibility(View.GONE);
                    fields.remove(passField);
                }
            });

            valueEdit.addTextChangedListener(new SimpleTextWatcher() {
                @Override
                public void afterTextChanged(final Editable s) {
                    passField.setValue(s.toString());
                }
            });

            labelEdit.addTextChangedListener(new SimpleTextWatcher() {
                @Override
                public void afterTextChanged(final Editable s) {
                    passField.setLabel(s.toString());
                }
            });
        }
    }

}

File: android/src/main/java/org/ligi/passandroid/ui/edit/FieldsEditFragment.kt
package org.ligi.passandroid.ui.edit

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import com.google.android.material.textfield.TextInputEditText
import org.koin.android.ext.android.inject
import org.ligi.kaxt.doAfterEdit
import org.ligi.passandroid.R
import org.ligi.passandroid.databinding.EditFieldsBinding
import org.ligi.passandroid.model.PassStore
import org.ligi.passandroid.model.pass.PassField
import org.ligi.passandroid.model.pass.PassImpl


class FieldsEditFragment : Fragment() {

    val passStore by inject<PassStore>()

    private fun getPass(): PassImpl = passStore.currentPass as PassImpl

    private var isEditingHiddenFields: Boolean = false

    private lateinit var inflater: LayoutInflater

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        this.inflater = inflater
        val inflate = EditFieldsBinding.inflate(layoutInflater, container, false)

        arguments?.let { isEditingHiddenFields = it.getBoolean(ARGUMENT_KEY) }

        if (isEditingHiddenFields) {
            inflate.addField.setText(R.string.add_back_fields)
        } else {
            inflate.addField.setText(R.string.add_front_field)
        }
        getPass().fields
            .filter { it.hide == isEditingHiddenFields }
            .forEach { addField(it, inflate.fieldsContainer) }

        inflate.addField.setOnClickListener {
            val passField = PassField(null, null, null, isEditingHiddenFields)
            addField(passField, inflate.fieldsContainer)
            getPass().fields.add(passField)
        }
        return inflate.root
    }

    private fun addField(passField: PassField, viewGroup: ViewGroup) {
        val child = inflater.inflate(R.layout.edit_field, viewGroup, false) as ViewGroup
        FieldView(child).apply(passField, getPass().fields)
        viewGroup.addView(child)
    }

    internal inner class FieldView(private val container: ViewGroup) {

        fun apply(passField: PassField, fields: MutableList<PassField>) {
            container.findViewById<TextInputEditText>(R.id.label_field_edit).apply {
                setText(passField.label)
                doAfterEdit {
                    passField.label = "$it"
                }
            }
            container.findViewById<TextInputEditText>(R.id.value_field_edit).apply {
                setText(passField.value)
                doAfterEdit {
                    passField.value = "$it"
                }
            }

            container.findViewById<View>(R.id.delete_button).setOnClickListener { v ->
                (v.parent as View).visibility = View.GONE
                fields.remove(passField)
            }

        }
    }

    companion object {

        private const val ARGUMENT_KEY = "KEY"

        fun create(primary: Boolean): FieldsEditFragment {
            val fieldsEditFragment = FieldsEditFragment()
            fieldsEditFragment.arguments = bundleOf(ARGUMENT_KEY to primary)
            return fieldsEditFragment
        }
    }

}
